p2p/src/network/identify/stream/p2p_network_identify_stream_reducer.rs

```diff
@@ -62,52 +115,173 @@ impl P2pNetworkIdentifyStreamState {
                     let data = &data[(data.len() - reader.len())..];
 
                     if len > reader.len() {
-                        *self = P2pNetworkIdentifyStreamState::IncomingPartialData {
+                        *stream_state = P2pNetworkIdentifyStreamState::IncomingPartialData {
                             len,
                             data: data.to_vec(),
                         };
                         Ok(())
                     } else {
-                        self.handle_incoming_identify_message(len, data)
+                        stream_state.handle_incoming_identify_message(len, data)?;
+                        if let P2pNetworkIdentifyStreamState::IdentifyReceived { data } =
+                            stream_state
+                        {
+                            let data = data.clone();
+                            let dispatcher = state_context.into_dispatcher();
```

Comment:
If possible I would obtain the dispatcher only once (instead of having a separate call per branch) to use it later. The idea is to have clear separations between state updates and validations, and control flow (dispatches) phases.

Reply:
Alternatively just a comment clearly separating the two sections (I see you still need to access the state at this point, but it is not going to be mutated anymore, so we are done with the state updates phase by now)